File: cpp-answers/169.多数元素.cpp
/*
 * @Author: your name
 * @Date: 2020-06-05 21:55:00
 * @LastEditTime: 2020-06-05 22:17:05
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \leetcode刷题\169.多数元素.cpp
 */
/*
 * @lc app=leetcode.cn id=169 lang=cpp
 *
 * [169] 多数元素
 */

// @lc code=start
class Solution
{
public:
    /*哈希表解法
    int majorityElement(vector<int> &nums)
    {
        unordered_map<int, int> counter;
        for (auto num : nums)
        {
            if (++counter[num] > nums.size() / 2)
            {
                return num;
            }
        }
        return 0;
    }*/
    /*nth_element方法
    int majorityElement(vector<int> &nums)
    {
        nth_element(nums.begin(), nums.begin() + nums.size() / 2, nums.end());
        return nums[nums.size() / 2];
    }*/
    /*Moore投票法*/
    int majorityElement(vector<int> &nums)
    {
        int counter = 0, majority;
        for (int num : nums)
        {
            if (!counter)
            {
                majority = num;
            }
            counter += num == majority ? 1 : -1;
        }
        return majority;
    }
};
// @lc code=end

File: rust-answers/169.多数元素.rs
/*
 * @lc app=leetcode.cn id=169 lang=rust
 *
 * [169] 多数元素
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut count = 1;

        for i in 1..nums.len() {
            if nums[res] == nums[i] {
                count += 1;
            } else {
                count -= 1;
            }
            if count == 0 {
                res = i;
                count = 1;
            }
        }

        count = 0;

        for i in 0..nums.len() {
            if nums[i] == nums[res] {
                count += 1;
            }
        }

        if count > (nums.len() / 2) {
            return nums[res];
        }
        return -1;
    }
}
// @lc code=end

